File: zerocool-kor/long-pressed-name.ts
import Debug from "debug";
const debug = Debug("long-pressed-name");

const isLongPressedName = (name:string, typed:string):boolean => {
  debug(`isLongPressedName started (name:${name}, typed:${typed})`);
  let isLongPressedNameIndex = 0;
  if(!name || !typed){
    return false;
  }
  const a = name.split('');
  const b = typed.split('');
  let pre;
  return !(b.find((c,i) => {
    debug(`[i:${i}] -> ${c}`);    
    if(a[0]===c){
      pre = a.shift();
    }else if(pre===c){
      /* OK */
    }else{
      return true;
    }
  }));
}

debug(`result -> ${isLongPressedName('aalex','aaleex')}`);
debug(`result -> ${isLongPressedName('saeed','ssaaedd')}`);
debug(`result -> ${isLongPressedName('leelee','llllleeelee')}`);
debug(`result -> ${isLongPressedName('laiden','laiden')}`);
debug(`result -> ${isLongPressedName('abbbbcd','aaaaaaaaaaabbbcd')}`);

File: zerocool-kor/leetcode/long-pressed-name.go
package main

import (
	"log"
	"strings"
)

func isLongPressedName(name string, typed string) bool {
	log.Println("isLongPressedName(name", name, ",typed:", typed, ") started")
	a := strings.Split(name, "")
	b := strings.Split(typed, "")
	log.Println("a (size:", len(a), ") -> ", a)
	log.Println("b (size:", len(b), ") -> ", b)
	pre := ""
	retval := true

	for i, c := range b {
		log.Println("[i:", i, "] -> ", c)
		if len(a) > 0 && a[0] == c {
			pre = a[0]
			a = a[1:]
		} else if pre == c {
			/* OK */
		} else {
			retval = false
			break
		}
	}

	return retval
}

func main() {
	log.Println("result -> ", isLongPressedName("aalex", "aaleex"))
	log.Println("result -> ", isLongPressedName("saeed", "ssaaedd"))
	log.Println("result -> ", isLongPressedName("leelee", "llllleeelee"))
	log.Println("result -> ", isLongPressedName("laiden", "laiden"))
	log.Println("result -> ", isLongPressedName("abbbbcd", "aaaaaaaaaaabbbcd"))
	log.Println("result -> ", isLongPressedName("aalex", "aaleexa"))
}
